todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/taskdetail/domain/LogicTest.kt

```diff
@@ -0,0 +1,91 @@
+package com.example.android.architecture.blueprints.todoapp.taskdetail.domain
+
+import com.example.android.architecture.blueprints.todoapp.data.Task
+import com.example.android.architecture.blueprints.todoapp.data.TaskDetails
+import com.spotify.mobius.test.NextMatchers.*
+import com.spotify.mobius.test.UpdateSpec
+import com.spotify.mobius.test.UpdateSpec.assertThatNext
+import org.junit.Test
+
+
+class LogicTest {
+    private val spec: UpdateSpec<Task, TaskDetailEvent, TaskDetailEffect> = UpdateSpec(::update)
+
+    @Test
+    fun completingATaskCompletesAndSavesIt() {
+        val activeTask: Task = createTask()
+        spec.given(activeTask)
+                .whenEvent(TaskDetailEvent.CompleteTaskRequested)
+                .then(
+                        assertThatNext(
+                                hasModel(activeTask.complete()),
+                                hasEffects(TaskDetailEffect.SaveTask(activeTask.complete()) as TaskDetailEffect))
+                )
+    }
+
+    @Test
+    fun activatingATaskUncompletesAndSavesIt() {
+        val details = TaskDetails("T1", "description", true)
+        val completedTask = Task("123", details)
+        spec.given(completedTask)
+                .whenEvent(TaskDetailEvent.ActivateTaskRequested)
+                .then(
+                        assertThatNext(
+                                hasModel(completedTask.activate()),
+                                hasEffects(TaskDetailEffect.SaveTask(completedTask.activate()) as TaskDetailEffect)))
```

Comment:
You can create some convenience functions here in the test
```
fun saveTask(t: Task) = TaskDetailEffect.SaveTask(t) as TaskDetailEffect
```
This will make your test look like
```
hasEffects(saveTask(completedTask.activate()))
```